### DISCUSSION TITLE
Retrieving data from a SAAS service

### DISCUSSION THREAD

**Discussion by johanjanssens** [joomlatools/joomlatools-pages#505] [Q&A] [answered]
I was not aware that Pages could do something like that so easily. Do you have a link to a kind of example in the doc?

_Note: Question was received on Joomla slack chit-chat, cross-posting it here._

**Comment by johanjanssens** [comment, marked as answer]
Retrieving data from a SAAS service with Pages is very easy. Pages handles data sources in a transparent way. It doesn't make a difference if you connect to a database, a webservice or a file on the filesystem. They all work in exactly the same way.

To work with data Pages uses [collections](https://github.com/joomlatools/joomlatools-pages/wiki/Collection) and specifically for webservices, there is a base [webservice collection]( https://github.com/joomlatools/joomlatools-pages/wiki/Web-Service-Collection) you can use.

Every collection can be:

- filtered
- sorted
- searched
- paginated

Pages offers out of the box support for JSON and can even read CSV from a webservice, which is a great way to integrate with things like Google Spreadsheets.

An example would be

```yaml
---
collection:
   model: webservice?url=http://example.com/api/endpoint
---
```
```php
<? foreach(collection() as $item) : ?>
   <?= $item->title ?>
<? endforeach; ?>
```

_Note: here I'm assuming that the service returns a `title` property_ 

The wiki documentation is not fully up to date yet with all the work done in the last year, but all the PR's are fully documented and worth checking out. Related to webservice collections following changes are made that are not in the wiki:

#### 1. Configure data root: https://github.com/joomlatools/joomlatools-pages/pull/333

Out of the box Pages offers transparent support for JSONAPI endpoints (https://jsonapi.org/) and is fully compatible with the spec. If the endpoint doesn't use JSONAPI you might need to manually specify where in the json the data can be found, this can be done using the `data`config option.

#### 2. Passing additional headers https://github.com/joomlatools/joomlatools-pages/pull/479

Sometimes your webservice requires authentication, or you need to pass additional headers. This can be easily done by specifying the headers needed in the frontmatter.

#### 3. Webservice caching: https://github.com/joomlatools/joomlatools-pages/pull/213

Pages implements a smart webservice cache. It knows how and when to refetch data from the webservice, this works out of the box, if needed you can set the cache time for a webservice collection manually through the `cache` config option

#### 4. Editable collections: https://github.com/joomlatools/joomlatools-pages/pull/189 (advanced)

Pages offers a full HTTP integration, you cannot only get data from an API endpoint you can also send data to it, this allows to create API's with ease. and it allows to turn a Joomla into a full fledged headless cms.

Technically this even allows you to create whole admin components with Pages. Do not have any examples of that yet, but it's very much possible and very simple.

#### 5. Collection API: https://github.com/joomlatools/joomlatools-pages/pull/144 (advanced)

The base collections in Pages try to cover 80% of the use-cases. In those case where the default implementation is not sufficient you can very easily create your own collections using the extensions API. An extension in pages is a simple file (containing a single class), you do not need to install it, just copy/paste it in place.

#### Related materials

Videos to watch

I did show some examples during my talk for Joomlashack last May already, and the talk of last Nov I tried to explain better what Pages is all about.

May:  https://www.youtube.com/watch?v=V13HaNKntuc
Nov: https://www.youtube.com/watch?v=srOb8VgEOmw

**Reply by woluweb** [reply]
I have not had the time to test/implement this already bc I first have to finish a project :)

But I was already thinking: based on your explanation above, we can retrieve data from a SASS service (let's say a list of Associations with their address, telephone number etc. A bit like you have on https://social.brussels/ which is a kind of directory of social services) in order to mix them with Joomla Articles.

Let me be more specific: can we easily mix on a page
- a native Joomla Article (_an Article called "Social Service blabla"_)
- with data coming from the API of that external website (_example : https://social.brussels/rest/organisation/13817_) 

If you have a little basic example then I can improve/customize on that.

Once more time, txs for Page!

PS: there will be a JoomlaDay USA next April (https://jdayusa.com/speakers). If you wish to present a few practical usecases that you have come across, it would help people to realize more quickly that this tool is the perfect match for many uses ;)

**Reply by johanjanssens** [reply]
No problem  at all, whenever you have time to test it and have questions happy to assist. About your follow-up, this is not hard at all and you might not even need a collection for it.

A collection can handle data in a dynamic way. It is like creating a listing page, collections can be searched, filtered, paginated and sorted using the URL or through formatter or the PHP API. In essence a collection creates a full REST API endpoint for your data. Pages knows what collections are used where and it can re-generate pages that use a collection when that collection's data changes. This makes collections very powerful.

99% of the data on a site will be accessed through a collection. However sometimes you just want to fetch a single piece of data and render that data somewhere. For example if you have a company site you want to show the address, contact info and opening hours on an about page. You could hardcode this data into HTML or in an article, but you probably need it in multiple places, rendered as html, but also rendered as micro-data etc. Here a collection would not make sense and hard-coding the data makes the site harder to maintain. 

#### Data API

To solve that there is the Data API: https://github.com/joomlatools/joomlatools-pages/wiki/Data The Data API is primarily used to work with static data stored in structured files on the file system. Data can be stored in JSON, XML, INI, YAML, and even supports PHP files that return an array.

For example: https://github.com/johanjanssens/agrea.ph/tree/master/sites/agrea.ph/data contains data in different formats that is rendered in diff places on http://agrea.ph/

The Data API can also fetch data through a URL. In your example all the code you would need is to render the name of the legal entity is: 

```php
<?= data('https://social.brussels/rest/organisation/13817')->legalStatus->labelNl; ?>
```

#### Related PR's

- https://github.com/joomlatools/joomlatools-pages/pull/264 shows how to use the data api to fetch a html page an extract html from it, in this case using a published Google Doc as an example but this works with any html page.

- https://github.com/joomlatools/joomlatools-pages/pull/512 improves the caching of remote data and gives you more control how long remote data is cached locally before being refetched.
